client/asset/eth/eth.go

```diff
@@ -2402,18 +2520,377 @@ func (eth *ETHWallet) checkForNewBlocks(reportErr func(error)) {
+		replacementTxHash, err := w.resubmitRedemption(tx, contractVer, nonceOverride, feeWallet)
+		if err != nil {
+			return nil, err
+		}
+
+		return confStatus(0, *replacementTxHash), nil
+	}
+
+	// Generally we rely on the transaction retrieved from geth, but we
+	// use our stored monitored tx for two reasons:
+	// - If the transaction was resubmitted, calls to ConfirmRedemption using
+	//   the old tx id will use the monitoredTxs map to determine the new one.
+	// - When geth cannot find the transaction, we can resubmit the redemption
+	//   using the tx stored in the monitoredTxs map.
+	w.monitoredTxsMtx.Lock()
+	monitoredTx, err := w.getLatestMonitoredTx(txHash)
+	if err != nil {
+		// We were unable to find the monitored tx, but we will attempt to continue
+		w.log.Error(err)
+	}
+	var blocksSinceSubmission int64
+	if monitoredTx != nil {
+		monitoredTx.mtx.Lock()
+		defer monitoredTx.mtx.Unlock()
+		if currentTip > monitoredTx.blockSubmitted {
+			blocksSinceSubmission = currentTip - monitoredTx.blockSubmitted
+		}
+		monitoredTxHash := monitoredTx.tx.Hash()
+		if monitoredTxHash != txHash {
+			w.log.Debugf("tx %s was replaced by %s since the last attempt to confirm redemption", txHash, monitoredTxHash)
+			txHash = monitoredTxHash
+		}
+	}
+	w.monitoredTxsMtx.Unlock()
+
+	gethTx, txBlock, err := w.node.getTransaction(w.ctx, txHash)
+	if errors.Is(err, asset.CoinNotFoundError) {
+		w.log.Errorf("ConfirmRedemption: geth could not find tx: %s", txHash)
+
+		if monitoredTx != nil && blocksSinceSubmission < blocksToWaitBeforeCoinNotFound {
+			return confStatus(0, txHash), nil
+		}
+
+		swapIsRedeemed, err := w.swapIsRedeemed(secretHash, contractVer)
+		if err != nil {
+			return nil, err
+		}
+		if swapIsRedeemed {
+			return confStatus(txConfsNeededToConfirm, txHash), nil
+		}
+
+		if monitoredTx != nil {
+			return resubmitTx(monitoredTx.tx, nil)
+		} else {
+			// If the tx is missing from both geth and the monitorTx, it is
+			// resubmitted individually.
+			txs, _, _, err := w.Redeem(&asset.RedeemForm{
+				Redemptions:   []*asset.Redemption{redemption},
+				FeeSuggestion: w.FeeRate(),
+			}, nil, nil)
+			if err != nil {
+				return nil, err
+			}
+			if len(txs) == 0 {
+				return nil, errors.New("no txs returned while resubmitting redemption")
+			}
+			var resubmittedTxHash common.Hash
+			copy(resubmittedTxHash[:], txs[0])
+			return confStatus(0, resubmittedTxHash), nil
+		}
+	} else if err != nil {
+		return nil, err
+	}
+
+	var confirmations int64
+	if txBlock > 0 {
+		if currentTip >= txBlock {
+			confirmations = currentTip - txBlock + 1
+		}
+	}
+
+	if confirmations >= txConfsNeededToConfirm {
+		receipt, err := w.node.transactionReceipt(w.ctx, txHash)
+		if err != nil {
+			return nil, err
+		}
+		if receipt.Status == types.ReceiptStatusSuccessful {
+			return confStatus(txConfsNeededToConfirm, txHash), nil
+		} else {
+			return resubmitTx(monitoredTx.tx, nil)
+		}
```

Comment:
> Also, the tx status checks also make sense below for any confs.

That's true.

We wouldn't get into a loop, because in `resubmitRedemption` it's checking `isRedeemable` for each of the swaps. If the transaction failed because one of them got refunded/redeemed earlier, then this will not be included in the new transaction.